framework/src/modules/dpos_v2/module.ts

```diff
@@ -130,21 +130,26 @@ export class DPoSModule extends BaseModule {
 
 	public async afterBlockExecute(context: BlockAfterExecuteContext): Promise<void> {
 		const { getStore, header } = context;
-
+		const isLastBlockOfRound = this._isLastBlockOfTheRound(header.height);
 		const previousTimestampStore = getStore(this.id, STORE_PREFIX_PREVIOUS_TIMESTAMP);
 		const previousTimestampData = await previousTimestampStore.getWithSchema<PreviousTimestampData>(
 			EMPTY_KEY,
 			previousTimestampStoreSchema,
 		);
 		const { timestamp: previousTimestamp } = previousTimestampData;
+
 		await this._updateProductivity(context, previousTimestamp);
+
+		if (isLastBlockOfRound) {
+			await this._createVoteWeightSnapshot(context);
+			await this._updateValidators(context);
```

Comment:
update validators should only happen after init round. `nextRoundNumber > initRounds`

Reply:
done